control/controlhttp/server.go

```diff
@@ -0,0 +1,90 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package controlhttp
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+
+	"tailscale.com/control/controlbase"
+	"tailscale.com/types/key"
+)
+
+// AcceptHTTP upgrades the HTTP request given by w and r into a
+// Tailscale control protocol base transport connection.
+//
+// AcceptHTTP always writes an HTTP response to w. The caller must not
+// attempt their own response after calling AcceptHTTP.
+func AcceptHTTP(ctx context.Context, w http.ResponseWriter, r *http.Request, private key.MachinePrivate) (*controlbase.Conn, error) {
+	next := r.Header.Get("Upgrade")
+	if next == "" {
+		http.Error(w, "missing next protocol", http.StatusBadRequest)
+		return nil, errors.New("no next protocol in HTTP request")
+	}
+	if next != upgradeHeader {
+		http.Error(w, "unknown next protocol", http.StatusBadRequest)
+		return nil, fmt.Errorf("client requested unhandled next protocol %q", next)
+	}
+
+	init, err := io.ReadAll(r.Body)
+	if err != nil {
+		http.Error(w, "error reading request", http.StatusInternalServerError)
+		return nil, fmt.Errorf("reading client request body: %v", err)
+	}
+
+	hijacker, ok := w.(http.Hijacker)
+	if !ok {
+		http.Error(w, "make request over HTTP/1", http.StatusBadRequest)
+		return nil, errors.New("can't hijack client connection")
+	}
+
+	w.Header().Set("Upgrade", upgradeHeader)
+	w.Header().Set("Connection", "upgrade")
+	w.WriteHeader(http.StatusSwitchingProtocols)
+
+	conn, brw, err := hijacker.Hijack()
+	if err != nil {
+		return nil, fmt.Errorf("hijacking client connection: %w", err)
+	}
+	if err := brw.Flush(); err != nil {
+		conn.Close()
+		return nil, fmt.Errorf("flushing hijacked HTTP buffer: %w", err)
+	}
+	if brw.Reader.Buffered() > 0 {
+		conn = &drainBufConn{conn, brw.Reader}
```

Comment:
I see... you're getting it from the HTTP body. That's a little fragile. I'd prefer to do what HTTP/2 does (and websockets kinda) and send the buffered beginning part base64 in a new HTTP header.

Reply:
(In any case, I'm very glad to be getting rid of the RTT. If we're going to have a single control region, we need to do everything we can to eliminate RTTs.)

Reply:
Can you say more about what's fragile about using the body? I find this surprising, so would like to document it for future readers.

Reply:
I was thinking of proxies that could read+forward only the headers and add "Expect: 100-continue". But I guess this would survive that.

Also if you've already sent a request body and the server replies with a non-2xx, the connection's basically dead at that point (out of sync) and can't be reused by the client. The server's probably going to close it immediately after sending the 2xx, since it doesn't want to deal with the rando request body for a request it rejected.

But if we're not ever going to send a non-2xx to these, maybe that's fine too.

I guess it's fine. It just weirds me out a bit.

Reply:
Yeah, for auth cases we get all the creds we need out of band. And for this particular protocol, we're only using HTTP as a stepping stone where we fully expect to either successfully stop being HTTP, or torch the entire connection on failure. Still happy to switch to headers if that makes you less nervous, but it sounds like we're okay with using the body?...